File: Year 2/Semester 2/MPP/JavaSocket/AirlinePersistence/src/main/java/airline/persistence/FlightRepository.java
package airline.persistence;


import airline.model.Flight;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;

import java.util.List;
import java.util.Properties;

public class FlightRepository implements IFlightRepository {
    private static final Logger logger= LogManager.getLogger();
    private JdbcUtils dbUtils;

    public FlightRepository(Properties properties) {
        dbUtils=new JdbcUtils(properties);
    }
    @Override
    public Iterable<Flight> getAll() {
        //logger.traceEntry();
        Connection connection= dbUtils.getConnection();
        List<Flight> flightList=new ArrayList<>();
        try(PreparedStatement ps=connection.prepareStatement("select * from flights")) {
            try (ResultSet resultSet = ps.executeQuery()) {
                while (resultSet.next()) {
                    Flight flight=extractEntity(resultSet);
                    flightList.add(flight);

                }
            }
        }
        catch(SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return flightList;
    }

    private Flight extractEntity(ResultSet resultSet) throws SQLException {
        int id=resultSet.getInt("flight_id");
        String destination=resultSet.getString("destination");
        String data=resultSet.getString("data");
        String hour=resultSet.getString("hour");
        int numberOfSeats=resultSet.getInt("no_seats");
        int availableSeats=resultSet.getInt("available_seats");
        Flight flight=new Flight(destination,data,hour,numberOfSeats);
        flight.setAvailableSeats(availableSeats);
        flight.setID(id);
        return flight;
    }

    @Override
    public Flight findById(int id) {
        logger.traceEntry();
        Connection connection= dbUtils.getConnection();
        try(PreparedStatement ps=connection.prepareStatement("select * from flights where id=?")){
            ps.setInt(1,id);
            try(ResultSet resultSet=ps.executeQuery()){
               resultSet.next();
               return extractEntity(resultSet);
            }
        }
        catch (SQLException ex){
            logger.error(ex);
        }

        logger.traceExit();
        return null;
    }

    @Override
    public void add(Flight entity) {
        logger.traceEntry();
        Connection connection= dbUtils.getConnection();
        try(PreparedStatement ps=connection.prepareStatement("insert into flights (data,hour,no_seats,destination,available_seats) values (?,?,?,?,)")){
            ps.setString(1,entity.getDate());
            ps.setString(2,entity.getHour());
            ps.setInt(3,entity.getNumberOfSeats());
            ps.setString(4,entity.getDestination());
            ps.setInt(5,entity.getAvailableSeats());
            int result=ps.executeUpdate();
            logger.trace("Added {} instances",result);
        }
        catch (SQLException ex){
            logger.error(ex);
        }
        logger.traceExit();
    }

    @Override
    public void updateNumberOfSeats(int flightId) {
        logger.traceEntry();
        Connection connection= dbUtils.getConnection();
        try(PreparedStatement ps=connection.prepareStatement("update flights set available_seats=available_seats-1 where flight_id=?")){
            ps.setInt(1,flightId);
            int result=ps.executeUpdate();
            logger.trace("Updated {} instances",result);
        }
        catch (SQLException ex){
            logger.error(ex);
        }
        logger.traceExit();
    }
}

File: Year 2/Semester 2/MPP/Protobuf/ProtoCSharp/AirlineServer/Persistency/FlightRepository.cs

using log4net;
using System;
using System.Collections.Generic;
using System.Data.Common;
using System.Data.SQLite;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Model;
using Persistency.utils;

namespace Persistency
{
    public class FlightRepository : IFlightRepository
    {
        private static readonly ILog log = LogManager.GetLogger("EmployeeRepository");
        private IDictionary<string, string> props;

        public FlightRepository(IDictionary<string, string> props)
        {
            this.props = props;
        }

        public void add(Flight entity)
        {
            log.Info("Entering add function");
            var connection = DBUtils.GetConnection(props);
            using (var command = new SQLiteCommand("insert into flights values (@id, @date, @hour, @no_seats, @destination, @available_seats)", connection as SQLiteConnection))
            {
                command.Parameters.AddWithValue("@id", null);
                command.Parameters.AddWithValue("@date", entity.Date);
                command.Parameters.AddWithValue("@available_seats", entity.AvailableSeats);
                var parameter = command.CreateParameter();
                parameter.ParameterName = "@hour";
                string chestie = entity.Hour;
                parameter.Value = entity.Hour;
                command.Parameters.Add(parameter);
                //command.Parameters.AddWithValue("@hour", entity.Hour);
                command.Parameters.AddWithValue("@no_seats", entity.NumberOfSeats);
                command.Parameters.AddWithValue("@destination", entity.Destination);
                command.ExecuteNonQuery();
            }
        }

        public void updateNumberOfSeats(int flightId)
        {
            var connection= DBUtils.GetConnection(props);
            using(var command=new SQLiteCommand("update flights set available_seats=available_seats-1 where flight_id=@flight_id",connection as SQLiteConnection))
            {
                command.Parameters.AddWithValue("@flight_id", flightId);
                command.ExecuteNonQuery();
            }
            
        }

        private Flight extractEntity(DbDataReader reader)
        {
            int id = reader.GetInt32(0);
            string destination = reader.GetString(4);
            string data = reader.GetString(1);
            string hour = reader.GetString(2);
            int no_seats = reader.GetInt32(3);
            int available_seats=reader.GetInt32(5);
            Flight flight = new Flight(destination, data, hour, no_seats,available_seats);
            flight.Id = id;
            return flight;

        }
        public Flight findById(int id)
        {
            log.InfoFormat("Entering findById with value {0}", id);
            var connection = (SQLiteConnection)DBUtils.GetConnection(props);
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select flight_id,data,hour,no_seats,destination,available_seats from flights where id=@id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);
                using (var reader = command.ExecuteReader())
                {
                    Flight flight = extractEntity(reader);
                    log.InfoFormat("Exiting findById with value {0}", flight);
                    return flight;
                }
            }
            log.InfoFormat("Exiting findById with value {0}", null);
            return null;
        }

        public List<Flight> GetAll()
        {
            var connection = (SQLiteConnection)DBUtils.GetConnection(props);
            List<Flight> flights = new List<Flight>();
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select flight_id,data,hour,no_seats,destination,available_seats from flights";
                using (var reader = command.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        Flight flight = extractEntity(reader);
                        flights.Add(flight);
                    }
                }
            }
            return flights;
        }
        
    }
}
